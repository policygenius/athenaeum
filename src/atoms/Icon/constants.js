module.exports = {
  icons: [
    'addAccount',
    'agent',
    'agent2',
    'americanExpress',
    'americanExpressSq',
    'americanExpressSquare',
    'animal',
    'application',
    'applicationColor',
    'armchair',
    'autoColor3X',
    'autogenius',
    'autogenius',
    'back',
    'backRebrand',
    'badWeather',
    'bag',
    'bandagedFinger',
    'bandagedHead',
    'basicChair',
    'bbb',
    'birthdayHat',
    'boneJar',
    'bookSearch',
    'books',
    'calculator',
    'calculatorGrey',
    'calculatorOrange',
    'calendar',
    'camera',
    'car',
    'carrier',
    'cart',
    'chat',
    'chatQuestion',
    'checkBlackGreyCircle',
    'checkCircle',
    'checkMark-1',
    'checklist',
    'checklistFolded',
    'checkmark',
    'checkmark-thin',
    'checkmarkBlack',
    'chevronDown',
    'chevronLeft',
    'chevronRight',
    'chevronUp',
    'circlePlus',
    'clock',
    'close',
    'closeAccent',
    'computer',
    'computerColor',
    'concierge',
    'dashboard',
    'digicert',
    'disability',
    'disabilityColor3X',
    'discount',
    'discountBadge',
    'discountBadge-shadow',
    'discountBlue',
    'discountGreen',
    'discountGreenCircle',
    'discountPill',
    'discover',
    'discoverSquare',
    'documentSmColor',
    'docUploadColor',
    'dog-1',
    'dog',
    'dogBones',
    'dogBowl',
    'dogCone',
    'down',
    'downWhite',
    'earch',
    'earchRebrand',
    'earthquake',
    'ellipsis',
    'email',
    'emailColor',
    'emailLock',
    'excellentFace',
    'excellentFaceGreen',
    'eyePatch',
    'facebook',
    'fallingObjects',
    'fastTrack',
    'family',
    'fastClock',
    'finalExpense',
    'fineArt',
    'fireExtinguisher',
    'firearm',
    'flood',
    'forYou',
    'foundation',
    'fourBedroom',
    'furCoat',
    'furCoatGray',
    'golfEquipment',
    'goodFace',
    'goodFaceGrey',
    'googlePlus',
    'greenCar',
    'greenCheckWithCircle',
    'greenCheckmark',
    'greenCircleDown',
    'greenCirclePlus',
    'guaranteedIssue',
    'hamburger',
    'hamburgerRebrand',
    'headAside',
    'health-1',
    'health',
    'healthColor3X',
    'healthPlan',
    'healthgenius',
    'healthgenius',
    'heart',
    'hmoPpo',
    'home',
    'homeAutoBundleColor',
    'homeBlack',
    'homeownersColor',
    'homeWhite',
    'hsaFsa',
    'hurricane',
    'idTheft',
    'ignOut',
    'ilverware',
    'ilverwareGray',
    'infinity',
    'instruments',
    'insurancegenius',
    'insurancegenius',
    'invite',
    'jewelry',
    'largeHome',
    'legacy',
    'lifeAgentsOff',
    'lifeAgentsOn',
    'life',
    'lifestyle',
    'lifeColor3X',
    'lightbulb',
    'linkedin',
    'lock',
    'lockBlue',
    'lockOrange',
    'luxurySofa',
    'madeInNy',
    'mastercard',
    'mastercardSquare',
    'medExam',
    'medicalExam',
    'medicalPayments',
    'medicalPaymentsBlue',
    'metlife',
    'moneyCloud',
    'moneygenius',
    'moneygenius',
    'more',
    'moreColor',
    'noMedExamNeededGreen',
    'noMedicalExamNeeded',
    'norton',
    'nortonSecured',
    'nortonW',
    'notesSimple',
    'openBook',
    'overflow',
    'pawprint',
    'pencil',
    'personalLiability',
    'pet',
    'petColor3X',
    'petgenius',
    'petgenius',
    'pgLogo',
    'pgLogoBlack',
    'pgLogoOrange',
    'phone',
    'phoneRebrand',
    'pinterest',
    'plus',
    'policygeniusSymbol',
    'policygeniusSymbolCenter',
    'priceSticker',
    'priceTag',
    'questionBlackGreyCircle',
    'questionMark',
    'questionMarkWhite',
    'quoteCards',
    'quotes',
    'quotes2',
    'quotes3',
    'quotesColor',
    'rclLogo',
    'redCircleUp',
    'renters',
    'rentersColor3X',
    'returnOfPremium',
    'scales-1',
    'scales',
    'search',
    'searchRebrand',
    'shoppingCart',
    'signOut',
    'silverware',
    'silverwareGray',
    'spousal',
    'star',
    'star2',
    'starLight',
    'steeringWheel',
    'stethoscope',
    'stillwaterLogoBlack',
    'stillwaterLogoWhite',
    'studio',
    'stuff',
    'tar',
    'techgenius',
    'techgenius',
    'theftProtection',
    'thumbsDown',
    'thumbsDownSelected',
    'thumbsUp-1',
    'thumbsUp',
    'thumbsUp2',
    'thumbsUpSelected',
    'tooltip',
    'tooltipBlack',
    'travelersColor3X',
    'tudio',
    'tuff',
    'tv',
    'twitter',
    'twoBedroom',
    'user',
    'usingTablet',
    'vandalizism',
    'veryPopular',
    'visa',
    'visaSquare',
    'visionColor3X',
    'waterDamage',
    'who',
    'xCircleSmBlack',
    'xIcon',
    'yellowTriangle',
  ]
};
