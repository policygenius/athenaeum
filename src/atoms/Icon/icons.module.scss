.icon-wrapper {
  position: relative;
  display: inline-block;
  width: rem-calc(72px);
  height: rem-calc(72px);
  display: inline-block;

  /* stylelint-disable selector-pseudo-class-no-unknown */
  // This class comes from our SVG dependency
  :global {
    .SVGInline-svg {
      width: 100%;
      height: 100%;
      display: block;
    }
  }
  /* stylelint-enable selector-pseudo-class-no-unknown */
}

.inline {
  vertical-align: middle;
  // This will get the icons to actually be centered inline.
  // SHOUTOUT: @dan.novograd for teaching the frontend team about 'ex' units
  transform: translateY(-0.27ex);
}

.left {
  margin-right: rem-calc(12);
}

.right {
  margin-left: rem-calc(12);
}

.clickable {
  cursor: pointer;
}
