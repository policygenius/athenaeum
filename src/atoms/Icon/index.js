import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import SVGInline from 'react-svg-inline';
import { icons as ICONS } from './constants';

import styles from './icons.module.scss';

const checkLS = () => typeof window === 'undefined' || !!window.lazySizes;

class Icon extends React.Component {
  constructor() {
    super();

    this.state = {
      svgString: null,
    };
  }

  componentDidMount() {
    const { renderSVGDOM, icon } = this.props;

    if (renderSVGDOM && icon) {
      this.getSVG();
    }
  }

  getSVG = () => {
    fetch(this.svgSrc)
      .then((data) => data.text())
      .then((svg) => {
        this.setState({
          svgString: svg,
        });
      });
  }

  get svgSrc() {
    return `https://static.policygenius.com/svgs/${this.sanitizedIcon}.svgz`;
  }

  get sanitizedIcon() {
    const { icon } = this.props;

    return icon.replace(/[^a-zA-Z0-9_-]/, '');
  }

  render() {
    const {
      className,
      onClick,
      height,
      width,
      inline,
      icon,
      lazy,
      alt,
      title,
      renderSVGDOM,
      ...rest
    } = this.props;

    const { svgString } = this.state;

    if (!icon) { return null; }

    const wrapperProps = {
      className: cx(
        styles['icon-wrapper'],
        onClick && styles.clickable,
        inline && styles.inline,
        inline && styles[inline],
        className
      ),
      onClick,
      style: {
        width,
        height
      },
      ...rest
    };

    if (renderSVGDOM && svgString) {
      return (
        <span {...wrapperProps}>
          <SVGInline
            svg={svgString}
          />
        </span>
      );
    }

    const isLazy = lazy && checkLS();
    const { sanitizedIcon } = this;
    const src = this.svgSrc;

    // lazyload css class comes from Img component
    return (
      <span {...wrapperProps}>
        <img
          className={cx(
            styles.img,
            isLazy && 'lazyload'
          )}
          alt={alt != null ? alt : sanitizedIcon}
          title={title || sanitizedIcon}
          data-src={isLazy ? src : undefined}
          src={isLazy ? undefined : src}
          {...rest}
        />
      </span>
    );
  }
}

Icon.propTypes = {

  /**
   * Supply any additional class names. Resize the icon via the parent's stylesheet.
   */
  className: PropTypes.string,

  /**
   * The icon you'd like to display from our cloud library.
   */
  icon: PropTypes.oneOf(ICONS).isRequired,

  /**
   * Click handler for the wrapper div around the svg
   */
  onClick: PropTypes.func,

  /**
   * Override the default height of the icon. Icon will maintain its proportions.
   */
  height: PropTypes.string,

  /**
   * Override the default width of the icon. Icon will maintain its proportions.
   */
  width: PropTypes.string,

  /**
   * For use with another component e.g. Text. Vertically aligns to the middle, adds margin.
   */
  inline: PropTypes.oneOfType([
    PropTypes.oneOf([ 'left', 'right' ]),
    PropTypes.bool
  ]),

  /**
   * Set to false to turn off lazyloading.
   */
  lazy: PropTypes.bool,

  /**
   * Defaults to icon name.
   */
  alt: PropTypes.string,

  /**
   * Defaults to icon name.
   */
  title: PropTypes.string,

  /**
   * Renders full SVG DOM element for requested icon
   */
  renderSVGDOM: PropTypes.bool,
};

Icon.defaultProps = {
  lazy: true
};

export default Icon;
