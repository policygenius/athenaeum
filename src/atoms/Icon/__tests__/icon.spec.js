import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

import Icon from 'atoms/Icon';

describe('<Icon />', () => {
  describe('when no icon is given', () => {
    const { error } = console;

    beforeEach(() => {
      console.error = jest.fn();
    });

    afterEach(() => {
      console.error = error;
    });

    it('returns null', () => {
      const wrapper = shallow(<Icon />);

      expect(wrapper.type()).toEqual(null);
      expect(console.error).toBeCalled();
    });
  });

  describe('when passed an icon as a name string', () => {
    it('renders a regular img element', () => {
      const props = {
        icon: 'xIcon'
      };

      const wrapper = mount(<Icon {...props} />);

      expect(wrapper.find('img')).toHaveLength(1);
    });
  });

  it('matches snapshot', () => {
    const props = {
      className: 'some-class',
      onClick: jest.fn(),
      height: '100px',
      width: '200px',
      inline: 'left',
      icon: 'xIcon'
    };

    const actual = renderer.create(<Icon {...props} />).toJSON();

    expect(actual).toMatchSnapshot();
  });

  it('accepts an empty string for alt text', () => {
    const props = {
      icon: 'xIcon',
      alt: ''
    };

    const wrapper = mount(<Icon {...props} />);

    expect(wrapper.find('img').prop('alt')).toBe('');
  });
});
