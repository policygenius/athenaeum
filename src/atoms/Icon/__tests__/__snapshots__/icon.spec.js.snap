// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon /> matches snapshot 1`] = `
<span
  className="icon-wrapper clickable inline left some-class"
  onClick={[MockFunction]}
  style={
    Object {
      "height": "100px",
      "width": "200px",
    }
  }
>
  <img
    alt="xIcon"
    className="img"
    src="https://static.policygenius.com/svgs/xIcon.svgz"
    title="xIcon"
  />
</span>
`;
